bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/ModuleUtil.java

```diff
@@ -0,0 +1,578 @@
+                resolvedOutputs.put(referencedOutput.getName(), outputValue);
+            } else {
+                getLog().error(
+                        "Output: " + referencedOutput.getName() + " value type " + outputValue.getClass().getName()
+                                + " is not assignableFrom referred Output type " + referencedOutput.getType());
+            }
+
+        }
+    }
+
+    /**
+     * Gets default value recursively.
+     *
+     * @param configDescription the source ConfigDescriptionParameter
+     * @param defaultValue the default value of the config property
+     * @param configContextMap map key: name of ConfigDescriptionParameter, value: the ConfigDescriptionParameter
+     * @param configuration map with config properties
+     * @return the default config value for passed configDescription
+     */
+    private static Object getConfigDefaultValue(ConfigDescriptionParameter configDescription,
+            Map<String, ConfigDescriptionParameter> configContextMap, Map<String, ?> configuration) {
+
+        String configDefaultValue = configDescription.getDefault();
+        Object resultValue = null;
+        while (resultValue == null && isParsable(configDefaultValue)) {
+            String referredConfigName = parse(configDefaultValue);
+            validateConfigType(configDescription, referredConfigName, configContextMap);
+            resultValue = configuration.get(referredConfigName);
+            configDefaultValue = configContextMap.get(referredConfigName).getDefault();
+        }
+        if (resultValue == null) {
+            try {
+                resultValue = convertToConfigType(configDefaultValue, configContextMap.get(configDescription.getName())
+                        .getType());
+            } catch (Exception e) {
+                getLog().error(e.getMessage(), e);
+            }
+        }
+        return resultValue;
+    }
+
+    /**
+     * Tries to convert raw value to its type.
+     *
+     * @param value raw value
+     * @param type the type to be converted to
+     * @return the converted value
+     * @throws Exception if conversion failed
+     */
+    private static Object convertToConfigType(String value, Type type) throws Exception {
+        Object result;
+        switch (type) {
+            case TEXT:
+                result = value;
+                break;
+            case BOOLEAN:
+                String trimmed = value.trim();
+                if (trimmed.equalsIgnoreCase("true")) {
+                    result = Boolean.TRUE;
+                } else if (trimmed.equalsIgnoreCase("false")) {
+                    result = Boolean.FALSE;
+                } else {
+                    result = null;
+                }
+                break;
+            case INTEGER:
+                result = Integer.valueOf(value);
+                break;
+            case DECIMAL:
+                result = Double.valueOf(value);
```

Comment:
we usually use BigDecimal for numbers (INTEGER+DECIMAL) in ConfigDescriptions to have full precision and no rounding effects, see https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/ThingFactory.java#L203. We could think about moving this method to ConfigDescriptionParameter (which already uses BigDecimal for min/max/step). Would that be ok for you?